packages/client/src/generation/utils.ts

```diff
@@ -344,3 +349,73 @@ export function unique<T>(arr: T[]): T[] {
 
   return result
 }
+
+/**
+ * Determines if arg types need generic <$PrismaModel> argument added.
+ * Essentially, performs breadth-first search for any fieldRefTypes that
+ * do not have corresponding `meta.source` defined.
+ *
+ * @param type
+ * @returns
+ */
+export function needsGenericModelArg(type: DMMF.SchemaArgInputType) {
```

Comment:
Maybe i don't see the larger picture here, but could this be simplified down into something along the lines of: 

```ts
export function needsGenericModelArg(type: DMMF.SchemaArgInputType) {
  if (type.location === 'fieldRefTypes') {
    return true
  }

  const hasFields =
    type.location === 'inputObjectTypes' && typeof type.type === 'object' && !type.type.meta?.source

  if (hasFields) {
    return type.type.fields.some((type) => needsGenericModelArg(type))
  }

  return false
}
```

?

Reply:
This is very similar to what I've started with, but it has a problem: could be recursive, for example `TypeA` has field of `TypeB` and type has a field of `TypeA`. In that case, this version will never terminate. We need to keep track of types we've already seen to avoid this. Current version is more or less standard approach to breadth- or depth-first search in the graph.

Reply:
> This is very similar to what I've started with, but it has a problem: types could be recursive, for example `TypeA` has field of `TypeB` and type has a field of `TypeA`. In that case, this version will never terminate. We need to keep track of types we've already seen to avoid this. Current version is more or less standard approach to breadth- or depth-first search in the graph.

Ok, I understand and this is why you have the 'visited' map. Nice work!